### ISSUE TITLE
Calling AzureBlobFileSystem.cat on a file path, adds a "/" to the end of the file path

### ISSUE THREAD

**Issue by brunorpinho** [dask/adlfs#214]
I was using dask + zarr to store arrays on azure blob storage using this library, but I ran into an issue. 

What happens is that:

1. At some point when loading a zarr file from the storage `fsspec.mapping.FSMap` is created (with `.fs` being an `AzureBlobFileSystem` instance);
2. It calls `self.fs.cat(k)` where `k` is a string representing a file path on the blob storage: e.g. `my-blob/my-array/.zarray`; and `self.fs` is the AzureBlobFileSystem instance;
3.  `AzureBlobFileSystem.cat` calls `AzureBlobFileSystem._expand_path` at some point which runs this line:

https://github.com/dask/adlfs/blob/3874b3e536fe6b24c824ee096566c8620b623dfa/adlfs/spec.py#L1351

4. The thing returns: `my-blob/my-array/.zarray/` and later on the loading crashes with the following error:

```
ResourceNotFoundError: Operation returned an invalid status 'The specified blob does not exist.'
ErrorCode:BlobNotFound
```

---

I think the solution is just to make sure we don't add a "/" in `_expand_path`

**Comment by TomAugspurger** [comment]
I think that https://github.com/intake/filesystem_spec/commit/767388a2b87d5281d41d7031618fd733d70d903b#diff-02a7ab40d6026e60bf52f78d93fc7146e53bc2455a8cb4073838297feaa090fbL343 is the root cause (cc @martindurant and @
isidentical). Consider:

```python
In [1]: import adlfs

In [2]: fs = adlfs.AzureBlobFileSystem(account_name="daymeteuwest")

In [3]: fs.exists("daymet-zarr/daily/hi.zarr/time/0/")
Out[3]: False

In [4]: _ = fs.cat(["daymet-zarr/daily/hi.zarr/time/0"])

In [5]: fs.exists("daymet-zarr/daily/hi.zarr/time/0/")
Out[5]: True
```

Clearly *that* behavior is incorrect. The result of `.exists()` shouldn't change on whether something has been cached. But I don't know which one we're saying is officially correct. Does the path `daymet-zarr/daily/hi.zarr/time/0/` (trailing `/`) exist because the key `daymet-zarr/daily/hi.zarr/time/0` exists, or no?

**Comment by TomAugspurger** [comment]
cc @isidentical from https://github.com/intake/filesystem_spec/commit/767388a2b87d5281d41d7031618fd733d70d903b#diff-02a7ab40d6026e60bf52f78d93fc7146e53bc2455a8cb4073838297feaa090fbL343 as well.

**Comment by isidentical** [comment]
If I am not missing something, the change on the fsspec upstream ensures that we won't get a `KeyError` on cases like this (since the condition already ensures that the version without trailing slash is in the cache, so that we should retrive the entry that we checked for in the condition). Maybe this might be because somewhere in the adlfs the cache is populated with entries that ends with trailing slashes (as per https://github.com/intake/filesystem_spec/issues/554#issuecomment-790090351)?

**Comment by TomAugspurger** [comment]
> Maybe this might be because somewhere in the adlfs the cache is populated with entries that ends with trailing slashes (as per intake/filesystem_spec#554 (comment))?

I don't believe so, unless there's another cache I'm missing.

```python

In [2]: fs = adlfs.AzureBlobFileSystem(account_name="daymeteuwest")

In [3]: _ = fs.cat(["daymet-zarr/daily/hi.zarr/time/0"])

In [4]: fs.exists("daymet-zarr/daily/hi.zarr/time/0/")
Out[4]: True

In [5]: list(fs.dircache)
Out[5]: ['daymet-zarr/daily/hi.zarr/time', 'daymet-zarr/daily/hi.zarr/time/0']
```

**Comment by hayesgb** [comment]
This looks like a collision between the above two issues, which sparked [fsspec #562](https://github.com/intake/filesystem_spec/issues/562).  

One of the challenges (I at least) have seen handling pseudo-directories is tied up in this subject.  Definitely need to remove the trailing slash in adlfs, which crept in when trying to deal with pseudo-directories.  We've since updated mkdir to a noop, which eliminates the issue.

The question of the KeyError for caching is probably a separate discussion since it can create edge cases like #190

**Comment by TomAugspurger** [comment]
Yeah. It'd be great to handle the "directory" / trailing `/` issue once and for all. Document that, and call it established behavior going forward. We need to decide on some invariants that we keep (like "`fs.exists(key)` implies `fs.open(key)` will work").

Spot checking GCS and AWS, both of them think that if you have a bucket with `/path/to/key`, but no `/path/to/key/`, then

```
>>> fs.exists("/path/to/key/")  # with trailing slash
True
```

And they will open the object at `/path/to/key` if you ask for `/path/to/key/`. So adlfs is the odd one out here.

**Comment by hayesgb** [comment]
I will push a fix today.

**Comment by TomAugspurger** [comment]
Thanks! I do think it's worth having a longer discussion around the ideal behavior here, about how to consistently handle pseduo-directories in these object stores. Maybe that discussion has happened though, I haven't followed closely.

I haven't tested it yet, but something like this might work

```diff
diff --git a/adlfs/spec.py b/adlfs/spec.py
index 298fa7b..b13bcfd 100644
--- a/adlfs/spec.py
+++ b/adlfs/spec.py
@@ -1263,6 +1263,11 @@ class AzureBlobFileSystem(AsyncFileSystem):
 
         async with self.service_client.get_blob_client(container_name, path) as bc:
             exists = await bc.exists()
+
+        if path.endswith("/") and not exists:
+            async with self.service_client.get_blob_client(container_name, path[:-1]) as bc:
+                exists = await bc.exists()
+
         return exists
 
     async def _pipe_file(self, path, value, overwrite=True, **kwargs):
```

That solves the `exists()` side. Unfortunately things like `.open()` will need to be updated as well...

**Comment by hayesgb** [comment]
Fixed with #217
